### DISCUSSION TITLE
How to filter tags with 'OR' logic over API?

### DISCUSSION THREAD

**Discussion by AndriusV4** [netbox-community/netbox#11703] [Q&A]
Hello,

I've been researching this rather simple use case, but cannot find the correct answer anywhere yet - what is the correct syntax of filtering objects over API by their tags using 'OR' logic? 

My use case is rather simple - **need an API call that returns all objects that have either 'foo' or 'bar' or both tags**:

As documentation states `&` will work as AND operator for ManytoMany fields such as tags. Nothing is mentioned about OR logic for such fields.
> For example, GET /api/dcim/sites/?tag=foo&tag=bar will return only sites which have both the "foo" and "bar" tags applied.

In this comment https://github.com/netbox-community/netbox/issues/3038#issuecomment-489737057 it was mentioned that ManyToManyFields should use `&` for OR and `,` for AND, but I am not sure if this was implemented since even the documentation states otherwise. And `,` doesn't work for me at all. For example the call `GET /api/dcim/sites/?tag=foo,bar` treats 'foo,bar' as one string and returns nothing.

I am mostly likely missing something really simple here, so any help would be greatly appreciated!

Tested on a local deployment with 3.4.3 and https://demo.netbox.dev/ (v3.4.4)

Thanks

**Comment by Kludge-As-A-Service** [comment]
hello, can't answer to your question directly, but I believe that the tagging management has been significantly reworked in a recent-ish release (can't find the exact one), so maybe don't take info from several years ago at face value on that matter, as some things may work differently now.

**Reply by AndriusV4** [reply]
I have checked all releases from 3.2 and weren't able anything related to tag or any ManytoMany field API filtering mention unfortunately.
I am not taking the 3yo comment on an issue as actual correct implementation, it was just the only information I found on this topic overall, so I tried to test it and it did not work as expected.
So just trying to get any help from the community if anybody has tried tag OR filtering over API, because all of my research and testing are not giving me any answers if this should even work at all.

**Reply by candlerb** [reply]
I'm afraid you're right:

* `status=active&status=offline` will show you devices with status active OR status offline
* `tag=foo&tag=bar` will only show you devices with tag foo AND tag bar

That's just how it is. You'll need to do two queries, one for `tag=foo` and one for `tag=bar`, and merge the results client-side.  (Possibly the second query could be `tag__n=foo&tag=bar`, but I don't know if `tag__n=foo` means "does not have any tag foo", or "has any tag which is not foo".  Like most REST APIs, the semantics of query strings are not well documented).

I've tested `tag=foo,bar` and it doesn't work (with Netbox v3.4.4)

I've looked through the source:
* `netbox/netbox/filtersets.py`: NetBoxModelFilterSet
* `netbox/extras/filters.py`: TagFilter
* this in turn calls [django_filters.ModelMultipleChoiceFilter](https://django-filter.readthedocs.io/en/stable/ref/filters.html#multiplechoicefilter) (annoyingly the package appears to be called "django-filter" but you import "django_filters")

and I don't see anything that might split on comma for "OR" logic.

**Reply by AndriusV4** [reply]
Thanks @candlerb, that's one of the answers I was looking for. Github is not allowing me mark your reply as the correct answer though, since it's a reply and not answer to the direct question I assume

**Reply by candlerb** [reply]
Also note that TagFilter sets `conjoined: True` as a default (meaning "AND"), and this is not overridden in NetBoxModelFilterSet, where it simply calls `tag = TagFilter()`.
